### PR TITLE
[all] Fix some C4661 warnings

### PR DESCRIPTION
**Pull request by epernod**
Warning specific to windows fixed for function specialization (class templated, set the specialization in cpp), fix some headers (Mapping & MultiMapping)

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-scene-tests]

File: SofaKernel/framework/sofa/core/MultiMapping.h [hunk 1]
```diff
@@ -287,49 +287,55 @@ class MultiMapping : public BaseMapping
...
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Vec3fTypes, sofa::defaulttype::Vec2fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Vec3fTypes, sofa::defaulttype::Vec1fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Vec6fTypes, sofa::defaulttype::Vec1fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Rigid3fTypes, sofa::defaulttype::Vec1fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Rigid3fTypes, sofa::defaulttype::Vec3fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Rigid3fTypes, sofa::defaulttype::Vec6fTypes >;
+template class SOFA_CORE_API MultiMapping< sofa::defaulttype::Rigid3fTypes, sofa::defaulttype::Rigid3fTypes >;
 #endif
 
 #ifndef SOFA_FLOAT
```
Comment (damienmarchal) [inline SofaKernel/framework/sofa/core/MultiMapping.h]: C'est bien plus facile d'utiliser les macro SOFA_WITH_FLOAT et SOFA_WITH_DOUBLE que les double négation du #ifndef SOFA_FLOAT 

Eg:
#ifdef SOFA_WITH_FLOAT
.... blabla avec float
#endif 

#ifdef SOFA_WITH_DOUBLE
.... blabla avec float
#endif

File: applications/plugins/Compliant/compliance/DampingCompliance.h [hunk 2]
```diff
@@ -114,9 +114,22 @@ class SOFA_Compliant_API DampingCompliance : public core::behavior::ForceField<T
 };
 
 
+#if defined(SOFA_EXTERN_TEMPLATE) && !defined(SOFA_COMPONENT_COMPLIANCE_DAMPINGCOMPLIANCE_CPP)
```
Comment (damienmarchal) [inline applications/plugins/Compliant/compliance/DampingCompliance.h:117]: virer le SOFA_EXTERN_TEMPLATE ...on utilise des template systématiquement, le test n'est plus pertinant depuis c++x11 je pense.

File: applications/plugins/Compliant/compliance/DampingCompliance.h [hunk 3]
```diff
@@ -114,9 +114,22 @@ class SOFA_Compliant_API DampingCompliance : public core::behavior::ForceField<T
...
+#ifndef SOFA_FLOAT
```
Comment (damienmarchal) [inline applications/plugins/Compliant/compliance/DampingCompliance.h:118]: #ifdef SOFA_WITH_FLOAT
extern template class SOFA_Compliant_API DampingCompliance<sofa::defaulttype::Vec6fTypes>;
extern template class SOFA_Compliant_API DampingCompliance<sofa::defaulttype::Vec3fTypes>;
extern template class SOFA_Compliant_API DampingCompliance<sofa::defaulttype::Vec1fTypes>;
#endif 

C'est plus joli :)

File: SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyAlgorithms.h [hunk 4]
```diff
@@ -156,35 +156,36 @@ class TriangleSetTopologyAlgorithms : public EdgeSetTopologyAlgorithms<DataTypes
 };
 
 
+#if defined(SOFA_COMPONENT_TOPOLOGY_TRIANGLESETTOPOLOGYALGORITHMS_CPP)
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaBaseTopology/TriangleSetTopologyAlgorithms.h]: Le code dans le #if n'apparait dans l'unité de compilation que lorsque SOFA_COMPONENT_TOPOLOGY_TRIANGLESETTOPOLOGYALGORITHMS_CPP est définit (cad dans le .cpp)

**Pull request merged**